### PR TITLE
Convert purge logic to plugins

### PR DESCRIPTION
**Pull request by jhedstrom**
Addresses #100 and replaces #101.
## Remaining
- [x] Convert global settings to use plugins
- [x] Unit and kernel tests for the new plugins and services
- [x] `MessageCron` test passing once global settings are converted
## New UI

screen shot 2016-08-01 at 11 49 52 am

screen shot 2016-08-01 at 11 50 06 am

### CODE REVIEW

File: message.services.yml [hunk 1]
```diff
@@ -0,0 +1,7 @@
+services:
+  plugin.manager.message.purge:
+    class: Drupal\message\MessagePurgePluginManager
+    parent: default_plugin_manager
+  message.purger:
```
Comment (amitaibu) [inline message.services.yml]: That's a good idea. I thought all the logic would sit inside `MessagePurgePluginManager` including the actual purging, but this separation makes sense to me. 👍

**Comment by jhedstrom** [comment]
@amitaibu I think this is ready for review.

**Comment by amitaibu** [comment]
Thanks! Can you also please attach a few screenshots showing the UI - just makes it easier to later know what to look for.

File: message.module [hunk 2]
```diff
@@ -108,127 +108,11 @@ function message_entity_delete(EntityInterface $entity) {
...
-      $purge_messages += message_get_purgeable_by_template($purge_limit, $no_override_template_names, $purge_settings);
-    }
-  }
-
-  // Delete all gathered messages.
-  if (!empty($purge_messages)) {
-    \Drupal::entityManager()->getStorage('message')->delete($purge_messages);
-  }
+  /** @var \Drupal\message\MessagePurger $message_purger */
+  $message_purger = \Drupal::service('message.purger');
```
Comment (amitaibu) [inline message.module]: maybe we should call it something like `$message_purge_orchestrator`, just so the distinction is clearer between this and the `MessagePurge` plugins?

Comment (amitaibu) [inline message.module]: I mean, change the service and class names, not just the variable

File: src/Entity/MessageTemplate.php [hunk 3]
```diff
@@ -126,17 +126,12 @@ class MessageTemplate extends ConfigEntityBundleBase implements MessageTemplateI
...
-   *    of the message template. IF not set, the default purge settings defined in
-   *    the "Message settings" will apply.
-   * - 'days': Optional; Maximal message age in days. IF not set, the default
-   *    purge settings defined in the
-   *    "Message settings" will apply.
+   * Purge settings:
+   * - 'purge_override': TRUE or FALSE override the global behavior.
+   *    "Message settings" will apply. Defaults to FALSE.
+   * - 'purge_methods': An array of purge method plugin configuration, keyed by
+   *   the plugin_id. An empty array indicates no purge is enabled (although
```
Comment (amitaibu) [inline src/Entity/MessageTemplate.php]: plugin_id => plugin ID.

File: src/Form/MessageSettingsForm.php [hunk 4]
```diff
@@ -24,10 +25,19 @@ class MessageSettingsForm extends ConfigFormBase {
    * The entity manager object.
    *
    * @var \Drupal\Core\Entity\EntityManagerInterface
+   *
+   * @todo Use the entity type manager service.
```
Comment (amitaibu) [inline src/Form/MessageSettingsForm.php:29]: what is blocking this?

File: src/Form/MessageSettingsForm.php [hunk 5]
```diff
@@ -82,39 +93,23 @@ public function __construct(ConfigFactoryInterface $config_factory, EntityManage
   public function buildForm(array $form, FormStateInterface $form_state) {
     $config = $this->config('message.settings');
 
-    $form['purge'] = [
+    // Uses the same form keys as the MessageTemplateForm so that the purge
```
Comment (amitaibu) [inline src/Form/MessageSettingsForm.php:96]: 👍

File: src/Form/MessageTemplateForm.php [hunk 6]
```diff
@@ -10,8 +10,11 @@
 use Drupal\Core\Entity\EntityForm;
 use Drupal\Core\Entity\EntityTypeInterface;
 use Drupal\Core\Form\FormStateInterface;
+use Drupal\Core\Url;
 use Drupal\message\Entity\MessageTemplate;
 use Drupal\message\FormElement\MessageTemplateMultipleTextField;
+use Drupal\message\MessagePurgePluginManager;
+use Symfony\Component\DependencyInjection\ContainerInterface;
```
Comment (amitaibu) [inline src/Form/MessageTemplateForm.php:16]: is it being used? Adding coder would be a nice next step :)

File: src/Form/MessageTemplateForm.php [hunk 7]
```diff
@@ -35,10 +64,10 @@ public function form(array $form, FormStateInterface $form_state) {
     $template = $this->entity;
 
     $form['label'] = [
-      '#title' => t('Label'),
+      '#title' => $this->t('Label'),
```
Comment (amitaibu) [inline src/Form/MessageTemplateForm.php:66]: 👍

File: src/Plugin/MessagePurge/Days.php [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+use Drupal\message\MessagePurgeBase;
+use Drupal\message\MessageTemplateInterface;
+use Symfony\Component\DependencyInjection\ContainerInterface;
+use Symfony\Component\HttpFoundation\RequestStack;
+
+/**
+ * Delete messages older than certain days.
+ *
+ * @MessagePurge(
+ *  id = "days",
```
Comment (amitaibu) [inline src/Plugin/MessagePurge/Days.php]: I think there's a missing space here in the indentation

File: src/Plugin/MessagePurge/Days.php [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+   * {@inheritdoc}
+   */
+  public function buildConfigurationForm(array $form, FormStateInterface $form_state) {
+    $form['days'] = [
+      '#type' => 'number',
+      '#min' => 1,
+      '#title' => t('Messages older than'),
+      '#description' => t('Maximal message age in days.'),
+      '#default_value' => $this->configuration['days'],
+      '#tree' => FALSE,
```
Comment (amitaibu) [inline src/Plugin/MessagePurge/Days.php:78]: is #tree TRUE by default?

**Comment by amitaibu** [comment]
Looks great. I've added a few minor comments, and still need to do a final run on my local.

File: src/Form/MessageSettingsForm.php (hunk 4 above)
Comment (jhedstrom) [inline src/Form/MessageSettingsForm.php:29]: Nothing blocking, it just isn't related to this change necessarily.

File: src/Plugin/MessagePurge/Days.php (hunk 9 above)
Comment (jhedstrom) [inline src/Plugin/MessagePurge/Days.php:78]: No, but it's set to TRUE for the each plugin, so must be returned to false so the plugin form handler need not be aware of the exterior tree structure.

File: src/Form/MessageTemplateForm.php (hunk 6 above)
Comment (jhedstrom) [inline src/Form/MessageTemplateForm.php:16]: They are being used, but I found one that wasn't :)

PHPCS is running apparently, but not actually failing the build...

**Comment by jhedstrom** [comment]
@amitaibu I've addressed the comments above I think.

I updated the summary with screenshots of the UI.

**Comment by amitaibu** [comment]
Wonderful, thank you!

**Pull request merged**